File: app/src/main/java/uk/co/rogerstech/servalot/Servalot.java
/*
    Servalot - An inetd like multi-server
    Copyright (C) 2018  Andrew Rogers

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

package uk.co.rogerstech.servalot;

import java.io.File;

public class Servalot {

    public static void main(String[] args){

        // Set filesDir to ~/.servalot/files
        File filesDir = new File(System.getProperty("user.home"),".servalot");
        filesDir = new File(filesDir,"files");

        System.out.println("Files Directory: "+filesDir.getAbsolutePath());

        // Start service manager
        ServiceManager serviceManager = new ServiceManager(filesDir, new File(filesDir,"services.tsv"));
        // TODO: register NodeFactoryBuilders
        serviceManager.load();

        // Endless loop (service threads should be running)
        int cnt=0;
        while(true) {
            try {
                // Sleep for 10 seconds
                Thread.sleep(10000);
                cnt+=10;
                System.out.println(""+cnt);
	            
            } catch(InterruptedException e) {
            }
        }
    }
}

File: example-packages/ServalotGUI/www/Servalot.js
/**
 *
 * @licstart  The following is the entire license notice for the 
 *  JavaScript code in this page.
 *
 * Copyright (C) 2020  Andrew Rogers
 *
 *
 * The JavaScript code in this page is free software: you can
 * redistribute it and/or modify it under the terms of the GNU
 * General Public License (GNU GPL) as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option)
 * any later version.  The code is distributed WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.
 *
 * As additional permission under GNU GPL version 3 section 7, you
 * may distribute non-source (e.g., minimized or compacted) forms of
 * that code without the copy of the GNU GPL normally required by
 * section 4, provided you include this license notice and a URL
 * through which recipients can access the Corresponding Source.
 *
 * @licend  The above is the entire license notice
 * for the JavaScript code in this page.
 *
 */
 
var Servalot = function() {
    this.pending = {};
    if (typeof wvi !== 'undefined') {
        var that = this;
        wvi.response = function(json) { if(log)log("R:"+json); that.wv_response(JSON.parse(json)); };
    }
};

Servalot.prototype.command = function(obj, callback) {
    if (typeof wvi !== 'undefined') {
        this.wv_command( obj, callback );
    }
    else
    {
        this.ws_command( obj, callback );
    }
};

Servalot.prototype.ws_command = function(obj, callback) {
    var hostname = window.location.hostname;
    var ws = new WebSocket("ws://" + hostname + ":8800");
    var that = this;
    ws.onopen = function(e) {
        ws.send(JSON.stringify(obj));
    };
    ws.onmessage = function (e) {
        if( callback !== undefined ) {
            callback(JSON.parse(e.data));
        }
    };
};

Servalot.prototype.wv_command = function(obj, callback) {

    if( callback === undefined ) callback = function(){};

    // Find smallest number not already pending and use for source port
    let src = 0;
    while( this.pending[src] !== undefined ) src=src+1;

    // Store the callback
    this.pending[src] = callback;

    // Create a transport control packet
    tcp = {src: ""+src, dst: "0", data: obj}; // Use dst=0 for Servalot command server

    wvi.command(JSON.stringify(tcp));
};

Servalot.prototype.wv_response = function(tcp) {
    let dst=tcp.dst;
    if( dst !== undefined ) {

        // Check for the FIN flag
        var tcf = tcp.tcf;
        if( tcf !== undefined && tcf.includes("f") ) {
            delete this.pending[dst];
        }
        else {
            var cb=this.pending[dst];
            if( cb !== undefined ) cb(tcp.data);
        }
    }
};

